Replace CanGc::note() calls with arguments passed by callers

There are a lot of CanGc::note() calls in components/script: https://github.com/search?q=repo%3Aservo%2Fservo%20CanGc%3A%3Anote&type=code

This was introduced in #33144, but in general any use of CanGc::note() is an opportunity to make the code more precise by adding a CanGc argument that is passed by the caller instead (see #33386 for an example). This is a task that will take a very, very long time to do all at once, so it's fine to just focus on one caller at a time and propagate it upwards until it becomes too much work!

If you want to work on this task, just leave a comment linking to the specific call sites that you are going to focus on.

There is no need to write or run tests for changes related to this issue; it is strictly about making garbage collection hazards more visible when reading code, and if the code compiles then everything works!

in this case: 
```
error[E0308]: mismatched types
   --> components\script\dom\bindings\structuredclone.rs:251:16
    |
251 |     read: Some(read_callback),
    |           ---- ^^^^^^^^^^^^^ incorrect number of function parameters
    |           |
    |           arguments to this enum variant are incorrect
    |
    = note: expected fn pointer `unsafe extern "C" fn(*mut js::jsapi::JSContext, *mut js::jsapi::JSStructuredCloneReader, *const js::jsapi::CloneDataPolicy, u32, u32, *mut libc::c_void) -> _`
                  found fn item `unsafe extern "C" fn(*mut JSContext, *mut JSStructuredCloneReader, ..., ..., ..., ..., ...) -> ... {read_callback}`
help: the type constructed contains `unsafe extern "C" fn(*mut js::jsapi::JSContext, *mut js::jsapi::JSStructuredCloneReader, *const js::jsapi::CloneDataPolicy, u32, u32, *mut libc::c_void, script_runtime::CanGc) -> *mut js::jsapi::JSObject {dom::bindings::structuredclone::read_callback}` due to the type of the argument passed
   --> components\script\dom\bindings\structuredclone.rs:251:11
    |
251 |     read: Some(read_callback),
    |           ^^^^^-------------^
    |                |
    |                this argument influences the type of `Some`
note: tuple variant defined here
   --> /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23\library\core\src\option.rs:582:5
```

do I create a wrapper?

--- a/components/script/dom/bindings/codegen/Bindings.conf
+++ b/components/script/dom/bindings/codegen/Bindings.conf
@@ -235,6 +235,7 @@ DOMInterfaces = {
 
 'URL': {
     'weakReferenceable': True,
+    'canGc': ['Parse'],
 },
 
 'VRDisplay': {

--- a/components/script/dom/url.rs
+++ b/components/script/dom/url.rs
@@ -162,6 +162,7 @@ impl URLMethods for URL {
         global: &GlobalScope,
         url: USVString,
         base: Option<USVString>,
+        can_gc: CanGc,
     ) -> Option<DomRoot<URL>> {
         // Step 1: Let parsedURL be the result of running the API URL parser on url with base,
         // if given.
@@ -176,7 +177,7 @@ impl URLMethods for URL {
         // These steps are all handled while mapping the Result to an Option<ServoUrl>.
         // Regarding initialization, the same condition should apply here as stated in the comments
         // in Self::Constructor above - construct it on-demand inside `URL::SearchParams`.
-        Some(URL::new(global, None, parsed_url.ok()?, CanGc::note()))
+        Some(URL::new(global, None, parsed_url.ok()?, can_gc))
     }
 
     /// <https://w3c.github.io/FileAPI/#dfn-createObjectURL>

